sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSchemaInferenceSuite.scala

```diff
@@ -0,0 +1,200 @@
+      partitionColumnNames = partitionColumns,
+      properties = properties)
+  }
+
+  // Creates CatalogTablePartition instances for adding partitions of data to our test table.
+  private def hiveCatalogPartition(location: String, index: Int): CatalogTablePartition
+    = CatalogTablePartition(
+      spec = Map("partcol1" -> index.toString, "partcol2" -> index.toString),
+      storage = CatalogStorageFormat(
+        locationUri = Option(s"${location}/partCol1=$index/partCol2=$index/"),
+        inputFormat = Option("org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat"),
+        outputFormat = Option("org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat"),
+        serde = Option("org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"),
+        compressed = false,
+        properties = Map("serialization.format" -> "1")))
+
+  // Creates a case-sensitive external Hive table for testing schema inference options. Table
+  // will not have Spark-specific table properties set.
+  private def setupCaseSensitiveTable(
+      tableName: String,
+      dir: File): Unit = {
+    spark.range(NUM_RECORDS)
+      .selectExpr("id as fieldOne", "id as partCol1", "id as partCol2")
+      .write
+      .partitionBy("partCol1", "partCol2")
+      .mode("overwrite")
+      .parquet(dir.getAbsolutePath)
+
+    val client = externalCatalog.client
+
+    val catalogTable = hiveExternalCatalogTable(
+      tableName,
+      dir.getAbsolutePath,
+      lowercaseSchema,
+      Seq("partcol1", "partcol2"))
+    client.createTable(catalogTable, true)
+
+    val partitions = (0 until NUM_RECORDS).map(hiveCatalogPartition(dir.getAbsolutePath, _)).toSeq
+    client.createPartitions(DATABASE, tableName, partitions, true)
+
+    // Check that the table returned by HiveExternalCatalog has schemaPreservesCase set to false
+    // and that the raw table returned by the Hive client doesn't have and Spark SQL properties
+    // set (table needs to be obtained from client since HiveExternalCatalog filters these
+    // properties out).
+    assert(!externalCatalog.getTable(DATABASE, TEST_TABLE_NAME).schemaPreservesCase)
+    val rawTable = externalCatalog.client.getTable(DATABASE, TEST_TABLE_NAME)
+    assert(rawTable.properties.filterKeys(_.startsWith(SPARK_SQL_PREFIX)) == Map.empty)
+  }
+
+  // Create a test table used for a single unit test, with data stored in the specified directory.
+  private def withTestTable(dir: File)(f: File => Unit): Unit = {
+    setupCaseSensitiveTable(TEST_TABLE_NAME, dir)
+    try f(dir) finally spark.sql(s"DROP TABLE IF EXISTS $TEST_TABLE_NAME")
+  }
+
+  private val inferenceKey = SQLConf.HIVE_CASE_SENSITIVE_INFERENCE.key
+
+  test("Schema should be inferred and written to table properties when INFER_AND_SAVE is " +
+    "specified") {
+    withSQLConf(inferenceKey -> HiveCaseSensitiveInferenceMode.INFER_AND_SAVE.toString) {
+      withTempDir { dir =>
+        withTestTable(dir) { dir =>
```

Comment:
Since ```dir``` isn't being used in any of the tests I'm just going to roll it in to ```withTempDir()``` and change the type of the code block that function takes to ```() => Unit```